pkg/apis/serving/v1beta1/inference_service_status.go

```diff
@@ -565,50 +565,75 @@ func (ss *InferenceServiceStatus) SetModelFailureInfo(info *FailureInfo) bool {
 	return true
 }
 
-func (ss *InferenceServiceStatus) PropagateModelStatus(statusSpec ComponentStatusSpec, podList *v1.PodList, rawDeployment bool) {
+func (ss *InferenceServiceStatus) PropagateModelStatus(statusSpec ComponentStatusSpec, podList *v1.PodList, rawDeployment bool, serviceStatus *knservingv1.ServiceStatus) bool {
 	// Check at least one pod is running for the latest revision of inferenceservice
 	totalCopies := len(podList.Items)
 	if totalCopies == 0 {
-		ss.UpdateModelRevisionStates(Pending, totalCopies, nil)
-		return
-	}
-	// Update model state to 'Loaded' if inferenceservice status is ready.
-	// For serverless deployment, the latest created revision and the latest ready revision should be equal
-	if ss.IsReady() {
-		if rawDeployment {
-			ss.UpdateModelRevisionStates(Loaded, totalCopies, nil)
-			return
-		} else if statusSpec.LatestCreatedRevision == statusSpec.LatestReadyRevision {
-			ss.UpdateModelRevisionStates(Loaded, totalCopies, nil)
-			return
+		if !rawDeployment {
+			// Make sure we haven't scaled down to 0 because of an error
+			for _, knativeCond := range serviceStatus.Conditions {
+				if knativeCond.Status == "False" {
+					// If any of the knative statuses are False, the model failed
+					// Hopefully the lastFailureInfo already has the info we need, so we don't update it here
+					ss.UpdateModelRevisionStates(FailedToLoad, totalCopies, nil)
+					return true
+				}
+			}
```

Comment:
I'm not sure if I'm missing it, but I don't manage to see coverage for these lines in the units. Is it possible to add a unit for this case?

Reply:
Resolved